File: frontend/src/services/ShowtimeService.ts
import axios from "axios";
import { API_URL } from "../constants";
import { Seat } from "../types/Seat";
import { Seatmap } from "../types/Seatmap";

export class ShowtimeService {
	public static async getShowtimeDateById(id: number): Promise<string> {
		try {
			const response = await axios.get<string>(
				`${API_URL}/showtimes/${id}/date`
			);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	public static async getSeatmapById(id: number): Promise<Seatmap> {
		try {
			const response = await axios.get<Seatmap>(
				`${API_URL}/showtimes/${id}/seatmap`
			);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	public static async getSeatsByShowtimeId(
		showtimeId: number
	): Promise<Seat[]> {
		try {
			const response = await axios.get<Seat[]>(
				`${API_URL}/showtimes/${showtimeId}/seats`
			);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	public static async isSeatReserved(
		showtimeId: number,
		seatId: number
	): Promise<boolean> {
		try {
			const response = await axios.get<boolean>(
				`${API_URL}/showtimes/${showtimeId}/seats/${seatId}/reserved`
			);
			return response.data;
		} catch (error) {
			throw error;
		}
	}

	public static async setSeatReserved(
		showtimeId: number,
		seatId: number
	): Promise<void> {
		try {
			await axios.post(
				`${API_URL}/showtimes/${showtimeId}/seats/${seatId}/reserve`
			);
		} catch (error) {
			throw error;
		}
	}
}

File: backend/src/main/java/com/AcmeBuddy/backend/services/ShowtimeService.java
package com.AcmeBuddy.backend.services;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.AcmeBuddy.backend.DTO.SeatDTO;
import com.AcmeBuddy.backend.DTO.SeatmapDTO;
import com.AcmeBuddy.backend.Mapper.SeatMapper;
import com.AcmeBuddy.backend.Mapper.SeatmapMapper;
import com.AcmeBuddy.backend.entities.Seat;
import com.AcmeBuddy.backend.entities.SeatAvailability;
import com.AcmeBuddy.backend.entities.Seatmap;
import com.AcmeBuddy.backend.repositories.ISeatAvailabilityRepository;
import com.AcmeBuddy.backend.repositories.ISeatRepository;
import com.AcmeBuddy.backend.repositories.IShowtimeRepository;

import jakarta.transaction.Transactional;

@Service
public class ShowtimeService {

    private final IShowtimeRepository showtimeRepository;
    private final ISeatRepository seatRepository;
    private final ISeatAvailabilityRepository seatAvailabilityRepository;

    @Autowired
    public ShowtimeService(IShowtimeRepository showtimeRepository, ISeatRepository seatRepository, ISeatAvailabilityRepository seatAvailabilityRepository) {
        this.showtimeRepository = showtimeRepository;
        this.seatRepository = seatRepository;
        this.seatAvailabilityRepository = seatAvailabilityRepository;
    }

    public LocalDateTime getShowtimeDateById(long id) {
        return showtimeRepository.findById(id).get().getDateTime();
    }

    public SeatmapDTO getSeatmapByShowtimeId(long showtimeId) {
        return showtimeRepository.findById(showtimeId)
                .map(showtime -> {
                    Seatmap seatmap = showtime.getSeatmap();
                    if (seatmap == null) {
                        throw new IllegalArgumentException("Seatmap not found for Showtime with id: " + showtimeId);
                    }
                    return SeatmapMapper.toDTO(seatmap);
                })
                .orElseThrow(() -> new IllegalArgumentException("Showtime not found with id: " + showtimeId));
    }

    public List<SeatDTO> getSeatsByShowtimeId(long showtimeId) {
        return showtimeRepository.findById(showtimeId)
                .map(showtime -> {
                    int seatmapId = showtime.getSeatmap().getId();
                    List<Seat> seats = seatRepository.findBySeatmapId(seatmapId);
                    return seats.stream().map(SeatMapper::toDTO).toList();
                })
                .orElseThrow(() -> new IllegalArgumentException("Showtime not found with id: " + showtimeId));
    }

    public boolean isSeatReserved(long seatId, long showtimeId) {
        return seatAvailabilityRepository.findBySeatIdAndShowtimeId(seatId, showtimeId)
                .map(SeatAvailability::isReserved)
                .orElseThrow(() -> new IllegalArgumentException(
                "No SeatAvailability found for seat ID " + seatId + " and showtime ID " + showtimeId));
    }

    @Transactional
    public void setSeatReserved(long seatId, long showtimeId) {
        SeatAvailability seatAvailability = seatAvailabilityRepository.findBySeatIdAndShowtimeId(seatId, showtimeId)
                .orElseThrow(() -> new IllegalArgumentException(
                "No SeatAvailability found for seat ID " + seatId + " and showtime ID " + showtimeId));

        seatAvailability.reserve();
        seatAvailabilityRepository.save(seatAvailability);
    }
}
